### DISCUSSION TITLE
Table of Contents dots/line from text to page number

### DISCUSSION THREAD

**Discussion by wbittlehsl** [QuestPDF/QuestPDF#229] [Q&A] [answered]
I found the [Table of Contents](https://github.com/QuestPDF/QuestPDF/blob/main/QuestPDF.ReportSample/Layouts/TableOfContentsTemplate.cs) and it's working great.  Is there any way to add "..." or a line to connect the left aligned text to the right aligned number?

![image](https://user-images.githubusercontent.com/81644233/169950661-4a42e86e-fed3-4fde-85bd-ed3e7cc349e3.png)

**Comment by bennetbo** [comment]
I believe this is not available out of the box, but definitly achievable.
My attempt below which modifes the table of content template you linked (just for straight lines).
```csharp
//TableOfContentsTemplate Line 47
row.ConstantItem(25).Text($"{number}.");
row.RelativeItem().Row(nestedRow =>
{
    nestedRow.AutoItem().Text(locationName);
    nestedRow.RelativeItem().PaddingHorizontal(2).TranslateY(-2).BorderBottom(1);
});
//....
```

Now with dashed lines, which is a little more complicated, because i dont think theres a way you can draw dashed lines in QuestPDF currently without using SkiaSharp directly.
```csharp
//TableOfContentsTemplate Line 47
row.ConstantItem(25).Text($"{number}.");
row.RelativeItem().Row(nestedRow =>
{
    nestedRow.AutoItem().Text(locationName);
    nestedRow.RelativeItem().PaddingHorizontal(2).AlignBottom().TranslateY(-2).Height(1).Canvas((c, s) =>
    {
        using var paint = new SKPaint();
        paint.StrokeWidth = s.Height;
        paint.PathEffect = SKPathEffect.CreateDash(new float[] { 2, 2 }, 0);
        c.DrawLine(0, 0, s.Width, 0, paint);
    });
});
//....
```

Result:
![image](https://user-images.githubusercontent.com/53836821/170093771-e5613485-175b-47d9-8025-4ca2556b4c2f.png)

There is still one problem which i am not sure how to solve: The spacing between the right aligned text and the dots, because the width of the items is different on each line.
I tried with `row.AutoItem()` instead of `row.ConstantItem(150)` but had no success.

**Comment by MarcinZiabek** [comment]
@Bebo-Maker provided a correct solution. Please find below a small variation of the code that I have used to test:

```csharp
.Row(row =>
{
    row.ConstantItem(20).Text($"{number}.");
    row.AutoItem().Text(locationName);
    
    row.RelativeItem().PaddingHorizontal(2).AlignBottom().TranslateY(-3).Height(1).Canvas((canvas, space) =>
    {
        // best to statically cache
        using var paint = new SKPaint
        {
            StrokeWidth = space.Height,
            PathEffect = SKPathEffect.CreateDash(new float[] { 1, 3 }, 0)
        };
        
        canvas.DrawLine(0, 0, space.Width, 0, paint);
    });
    
    row.AutoItem().Text(text =>
    {
        text.BeginPageNumberOfSection(locationName);
        text.Span(" - ");
        text.EndPageNumberOfSection(locationName);
    });
});
```

That produces:
![image](https://user-images.githubusercontent.com/9263853/170137095-04796334-9557-42cc-b285-f002a4c31eb5.png)

**Important:** the current 2022.5 release will produce incorrect result due to the corner case. Short description of the internal issue: every document in QuestPDF is generated twice. First pass is to create overall virtual layout and gather information regarding page-number-related data, second pass renders target PDF file. To speed up the process, additional cache is being applied. One of the methods is that the Row element caches width of the AutoItem child. This is not an issue in most cases. However, the text, that shows page number, changes its width. Therefore,  the actual text width is different than the cached one.

I have already applied this fix to the main repository. As this is a lower-priority issue, I am planning to make it public with the 2022.6 release. Target date is June 7 2022. A prerelease should be available much earlier. I hope that it does not present much inconvinience.

**Reply by bennetbo** [reply]
Aha so thats why my attempt with `AutoItem` didnt work 😅.

**Reply by wbittlehsl** [reply]
@MarcinZiabek @Bebo-Maker Really appreciate the quick feedback!

@MarcinZiabek  the supplied code is giving me an "Object reference not set to an instance of an object." exception:

Here's the stack:
```
   at QuestPDF.Drawing.SkiaCanvasBase.Translate(Position vector)
   at QuestPDF.Elements.Column.Draw(Size availableSpace)
   at QuestPDF.Drawing.DocumentGenerator.RenderPass[TCanvas](PageContext pageContext, TCanvas canvas, Container content, DocumentMetadata documentMetadata, DebuggingState debuggingState)
```

My version of your code:
```csharp
//row.ConstantItem(20).Text($"{number}.");
row.AutoItem().Text(title);
row.RelativeItem().PaddingHorizontal(2).AlignBottom().TranslateY(-3).Height(1).Canvas((canvas, space) =>
{
    // best to statically cache
    using var paint = new SKPaint
    {
        StrokeWidth = space.Height,
        PathEffect = SKPathEffect.CreateDash(new float[] { 1, 3 }, 0)
    };
    canvas.DrawLine(0, 0, space.Width, 0, paint);
});
row.AutoItem().Text(text =>
{
    text.BeginPageNumberOfSection(destination);
    //text.Span(" - ");
    //text.EndPageNumberOfSection(locationName);
});
```

Any idea why I'd be getting this? I'm on 2022.5.0.

**Reply by MarcinZiabek** [reply]
Imagine my surprise when the obvious code did not work 😅

**Reply by MarcinZiabek** [reply]
@wbittlehsl This exception is strange, never had such an exception before. Just checked the implementation, so far I don't see any obvious place where `null` could be created. Would you like to provide more details? Is there an option to get access to your code?

**Reply by wbittlehsl** [reply]
@MarcinZiabek It may take me some time to put together a reproduction case that's outside my current project.  In the meantime I've noticed that if I change the last `row.AutoItem()` to `row.ConstantItem(30)` it works again.  Something about ending with an AutoItem produces the error for me.  Otherwise everything about the code above and my code are the same.

```csharp
//row.ConstantItem(20).Text($"{number}.");
row.ConstantItem(spacing);
row.AutoItem().Text(title);
row.RelativeItem().PaddingHorizontal(2).AlignBottom().TranslateY(-3).Height(1).Canvas((canvas, space) =>
{
    // best to statically cache
    using var paint = new SKPaint
    {
        StrokeWidth = space.Height,
        PathEffect = SKPathEffect.CreateDash(new float[] { 1, 3 }, 0)
    };
    canvas.DrawLine(0, 0, space.Width, 0, paint);
});
// this doesn't work
//row.AutoItem().Text(text =>
// this does work
row.ConstantItem(30).Text(text =>
{
    text.BeginPageNumberOfSection(destination);
    //text.Span(" - ");
    //text.EndPageNumberOfSection(locationName);
});
```
Debugging QuestPDF, this is happening on the second render pass which uses a `PdfCanvas`. The `PdfCanvas` itself is not null, but it's canvas, which is of type `SKCanvas`, IS null.  The `SKDocument` is NOT null.  The exception is thrown on page 2 of the document I'm generating.  Debugging `DocumentGenerator` line 106, the `BeginPage` method sets the SKCanvas to a null canvas.  The problem is that the `spacePlan` given to the `BeginPage` method has height/width both = 0.  This _appears_ to cause Skia to return a null canvas.  My page 2 is the Table of Contents page itself where the only content is the `TableOfContentsTemplate`.

**Reply by hisuwh** [reply]
Has anyone got a way to do this in later versions of QuestPDF where `.Canvas` is deprecated?

**Reply by BeneHenke** [reply]
You can use .Svg or try the dotted line https://www.questpdf.com/api-reference/line.html

**Comment by wbittlehsl** [comment]
@MarcinZiabek @Bebo-Maker It's me again... I'm having an issue with this sample code.  If the text in the TOC is too long I get a `DocumentLayoutException` because the `RelativeItem` (the ....) doesn't have space.

I've tried putting a `MaxWidth` on the `AutoItem` which resolves the issue, but the ..... is only rendered from the maximal box location (not from the wrapped text location).  For example:

```
This is a super long TOC title that gets wrapped because of the        45
MaxWidth setting                                                 ......
```

I've seen this from other libraries:
```
This is a super long TOC title that gets wrapped because of the
MaxWidth setting.......................................................45
```

But I'd be happy with this too (clipping the text instead of wrapping it):
```
This is a super long TOC title that gets wrapped because of the........45
```

**Reply by BeneHenke** [reply]
Did you find a way to do this?

**Comment by Kazagan** [comment]
Coming in with the solution I used for dotted lines between text: 
```
Row(r => 
{
     r.RelativeItem()
            .PaddingLeft(textPadding)
            .Layers(layer =>
            {
                string dots = new('.', 500); // string of 500 dots, to act as connector
                layer.Layer()
                    .AlignBottom()
                    .AlignRight()
                    .Text(dots)
                    .FontSize(_fontSize)
                    .ClampLines(1);
                layer.PrimaryLayer()
                    .AlignBottom()
                    .PaddingLeft(textPadding)
                    .Text(_left)
                    .FontSize(fontSize)
                    .BackgroundColor(Colors.White)
                    .ClampLines(_wrapLines);
            });
        r.AutoItem()
            .PaddingRight(textPadding)
            .AlignBottom()
            .AlignRight()
            .Background(Colors.White)
            .Text(_right)
            .FontSize(fontSize);
}
```
I needed dots, not dash, but I used periods, that stretched across the line, then applied a white background to the areas of the text I don't want them to appear. This worked well for my requirement.

**Comment by DawidRobertNowak** [comment, marked as answer]
Hello,

We are closing this thread due to inactivity. Thank you for taking the time to reach out and contribute to the project — your feedback helps us improve QuestPDF.

For anyone looking for an updated approach now that `.Canvas` has been deprecated, the [Line](https://www.questpdf.com/api-reference/line.html) element with `LineDashPattern()` offers a built-in way to draw dashed or dotted lines. Here is a simple example of a Table of Contents entry with a dotted leader:

```csharp
static void GenerateTableOfContentsSection(IContainer container)
{
    container
        .Column(column =>
        {
            column.Spacing(15);
            
            column
                .Item()
                .Text("Table of Contents")
                .Bold()
                .FontSize(20)
                .FontColor(Colors.Blue.Medium);
            
            column.Item()
                .Column(column =>
                {
                    column.Spacing(5);
                    
                    foreach (var i in Enumerable.Range(1, 10))
                    {
                        column.Item()
                            .SectionLink($"section-{i}")
                            .Row(row =>
                            {
                                row.ConstantItem(25).Text($"{i}.");
                                row.AutoItem().Text(Placeholders.Label());
                                row.RelativeItem().PaddingHorizontal(2).OffsetY(11).LineHorizontal(1).LineDashPattern([1, 3]);
                                row.AutoItem().Text(text => text.BeginPageNumberOfSection($"section-{i}"));
                            });
                    }
                });
        });
}
```

Should you encounter any further problems, please don't hesitate to open a new issue.

Thank you!
